src/algorithms/digi/PulseCombiner.cc

```diff
@@ -0,0 +1,153 @@
+      throw JException("Failed to load ID decoder");
+    }
+  }
+
+}
+
+void PulseCombiner::process(const PulseCombiner::Input& input,
+                            const PulseCombiner::Output& output) const {
+  const auto [inPulses] = input;
+  auto [outPulses]      = output;
+
+  // Create map containing vector of pulses from each CellID
+  std::map<uint64_t, std::vector<edm4hep::TimeSeries>> cell_pulses;
+  for (const edm4hep::TimeSeries& pulse: *inPulses) {
+    uint64_t shiftedCellID = pulse.getCellID() & m_detector_bitmask;
+    cell_pulses[shiftedCellID].push_back(pulse);
+  }
+
+  // Loop over detector elements and combine pulses
+  for (const auto& [cellID, pulses] : cell_pulses) {
+    if(pulses.size() == 1) {
+      outPulses->push_back(pulses.at(0).clone());
+      debug("CellID {} has only one pulse, no combination needed", cellID);
+    }
+    else{
+      std::vector<std::vector<edm4hep::TimeSeries>> clusters = clusterPulses(pulses);
+      for (auto cluster: clusters) {
+        // Clone the first pulse in the cluster
+        auto sum_pulse = outPulses->create();
+        sum_pulse.setCellID(cluster[0].getCellID());
+        sum_pulse.setInterval(cluster[0].getInterval());
+        sum_pulse.setTime(cluster[0].getTime());
+
+        auto newPulse = sumPulses(cluster);
+        for(auto pulse : newPulse) {
+          sum_pulse.addToAmplitude(pulse);
+        }
+      }
+      debug("CellID {} has {} pulses, combined into {} clusters", cellID, pulses.size(), clusters.size());
+    }
+
+  }
+
+} // PulseCombiner:process
+
+std::vector<std::vector<edm4hep::TimeSeries>> PulseCombiner::clusterPulses(const std::vector<edm4hep::TimeSeries> pulses) const {
+
+  //Clone the pulses aray of pointers so they aren't const
+  std::vector<edm4hep::TimeSeries> ordered_pulses;
+  for(auto pulse : pulses) {
+    ordered_pulses.push_back(pulse);
+  }
+
+  // Sort pulses by time, greaty simplifying the combination process
+  std::sort(ordered_pulses.begin(), ordered_pulses.end(), [](edm4hep::TimeSeries a, edm4hep::TimeSeries b) {
+    return a.getTime() < b.getTime();
+  });
+
+  // Create vector of pulses
+  std::vector<std::vector<edm4hep::TimeSeries>> cluster_pulses;
+  float clusterEndTime = 0;
+  bool isNewPulse = true;
+  // Create clusters of pulse indices which overlap with at least the minimum separation
+  for (auto pulse: ordered_pulses) {
+    float pulseStartTime = pulse.getTime();
+    float pulseEndTime = pulse.getTime() + pulse.getInterval()*pulse.getAmplitude().size();
+    if(!isNewPulse) {
+      if (pulseStartTime < clusterEndTime + m_minimum_separation) {
+        cluster_pulses.back().push_back(pulse);
+        clusterEndTime = std::max(clusterEndTime, pulseEndTime);
+      } else {
+        isNewPulse = true;
+      }
+    }
+    if(isNewPulse) {
+      cluster_pulses.push_back({pulse});
+      clusterEndTime = pulseEndTime;
+      isNewPulse = false;
```

Comment:
```suggestion
  bool makeNewPulse = true;
  // Create clusters of pulse indices which overlap with at least the minimum separation
  for (auto pulse: ordered_pulses) {
    float pulseStartTime = pulse.getTime();
    float pulseEndTime = pulse.getTime() + pulse.getInterval()*pulse.getAmplitude().size();
    if(!makeNewPulse) {
      if (pulseStartTime < clusterEndTime + m_minimum_separation) {
        cluster_pulses.back().push_back(pulse);
        clusterEndTime = std::max(clusterEndTime, pulseEndTime);
      } else {
        makeNewPulse = true;
      }
    }
    if(makeNewPulse) {
      cluster_pulses.push_back({pulse});
      clusterEndTime = pulseEndTime;
      makeNewPulse = false;
```
Maybe it's just me, but this name helps to better understand the code when reading from top to bottom.